File: python_project/src/static/indicators/nadaraya_watson.js
/* Nadaraya-Watson Envelope, based on the LuxAlgo Pine implementation.
 * The repainting variant uses future bars inside the latest window; the
 * endpoint variant only uses bars available at each point in time. */
(function () {
  const gaussian = (distance, bandwidth) =>
    Math.exp(-(distance * distance) / (2 * bandwidth * bandwidth));

  const envelope = (line, source, multiplier, start, end) => {
    const errors = [];
    for (let i = start; i < end; i += 1) {
      if (line[i] != null && source[i] != null) errors.push(Math.abs(source[i] - line[i]));
    }
    const mae = errors.length ? errors.reduce((sum, value) => sum + value, 0) / errors.length * multiplier : 0;
    return {
      line,
      upper: line.map(value => value == null ? null : value + mae),
      lower: line.map(value => value == null ? null : value - mae),
    };
  };

  function repainting(source, bandwidth = 8, multiplier = 3, lookback = 500) {
    const line = new Array(source.length).fill(null);
    const start = Math.max(0, source.length - lookback);
    for (let i = start; i < source.length; i += 1) {
      let weightedSum = 0;
      let weightSum = 0;
      for (let j = start; j < source.length; j += 1) {
        if (source[j] == null) continue;
        const weight = gaussian(i - j, bandwidth);
        weightedSum += source[j] * weight;
        weightSum += weight;
      }
      line[i] = weightSum ? weightedSum / weightSum : null;
    }
    return envelope(line, source, multiplier, start, source.length);
  }

  function endpoint(source, bandwidth = 8, multiplier = 3, lookback = 500) {
    const line = new Array(source.length).fill(null);
    const errors = new Array(source.length).fill(null);
    for (let i = lookback - 1; i < source.length; i += 1) {
      let weightedSum = 0;
      let weightSum = 0;
      for (let lag = 0; lag < lookback; lag += 1) {
        const value = source[i - lag];
        if (value == null) continue;
        const weight = gaussian(lag, bandwidth);
        weightedSum += value * weight;
        weightSum += weight;
      }
      line[i] = weightSum ? weightedSum / weightSum : null;
      errors[i] = Math.abs(source[i] - line[i]);
    }

    const upper = new Array(source.length).fill(null);
    const lower = new Array(source.length).fill(null);
    for (let i = lookback * 2 - 2; i < source.length; i += 1) {
      const mae = errors.slice(i - lookback + 1, i + 1)
        .reduce((sum, value) => sum + (value ?? 0), 0) / lookback * multiplier;
      upper[i] = line[i] + mae;
      lower[i] = line[i] - mae;
    }
    return { line, upper, lower };
  }

  window.NadarayaWatsonEnvelope = { repainting, endpoint };
}());

File: backtest/gold_nwe_sentiment/indicators/nadaraya_watson.py
"""Nadaraya-Watson Envelope, variante endpoint (no-repaint).

Porting fedele di static/indicators/nadaraya_watson.js (funzione `endpoint`) usata nel
grafico web, così la banda calcolata qui per il backtest è identica a quella mostrata
in UI — a differenza della variante "repainting", usa solo le barre già disponibili in
ogni istante, quindi è l'unica adatta a un backtest senza lookahead bias.
"""

import numpy as np


def _gaussian_weights(bandwidth: float, lookback: int) -> np.ndarray:
    lags = np.arange(lookback)
    return np.exp(-(lags ** 2) / (2 * bandwidth ** 2))


def endpoint_envelope(
    source: np.ndarray, bandwidth: float = 8, multiplier: float = 3, lookback: int = 500
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Ritorna (line, upper, lower). I primi 2*lookback-2 valori sono NaN (serve storico)."""
    n = len(source)
    line = np.full(n, np.nan)
    errors = np.full(n, np.nan)
    weights = _gaussian_weights(bandwidth, lookback)

    for i in range(lookback - 1, n):
        window = source[i - lookback + 1 : i + 1][::-1]  # lag 0 = barra i, lag k = i-k
        mask = ~np.isnan(window)
        if not mask.any():
            continue
        weight_sum = weights[mask].sum()
        if weight_sum == 0:
            continue
        line[i] = np.dot(window[mask], weights[mask]) / weight_sum
        errors[i] = abs(source[i] - line[i])

    upper = np.full(n, np.nan)
    lower = np.full(n, np.nan)
    band_start = lookback * 2 - 2
    for i in range(band_start, n):
        window_err = np.nan_to_num(errors[i - lookback + 1 : i + 1])
        mae = window_err.sum() / lookback * multiplier
        upper[i] = line[i] + mae
        lower[i] = line[i] - mae

    return line, upper, lower


def endpoint_line_errors_fast(source: np.ndarray, bandwidth: float, lookback: int = 500) -> tuple[np.ndarray, np.ndarray]:
    """Come la parte 'line/errors' di endpoint_envelope ma vettorizzata via convoluzione —
    stesso risultato numerico, molto più veloce quando bisogna provare molti bandwidth
    diversi (grid search fine). Richiede `source` senza NaN (vero per i prezzi Close)."""
    n = len(source)
    weights = _gaussian_weights(bandwidth, lookback)
    weight_sum = weights.sum()

    conv = np.convolve(source, weights, mode="valid") / weight_sum
    line = np.full(n, np.nan)
    line[lookback - 1 :] = conv

    errors = np.full(n, np.nan)
    errors[lookback - 1 :] = np.abs(source[lookback - 1 :] - line[lookback - 1 :])
    return line, errors


def bands_from_line_errors(line: np.ndarray, errors: np.ndarray, multiplier: float, lookback: int) -> tuple[np.ndarray, np.ndarray]:
    """Data una 'line/errors' già calcolata, applica un moltiplicatore per ottenere le bande
    — a costo quasi nullo, così si può riusare la stessa line per molti multiplier diversi."""
    n = len(line)
    err_filled = np.nan_to_num(errors)
    cumsum = np.cumsum(np.insert(err_filled, 0, 0.0))
    rolling_sum = cumsum[lookback:] - cumsum[:-lookback]
    mae = rolling_sum / lookback * multiplier

    upper = np.full(n, np.nan)
    lower = np.full(n, np.nan)
    valid_from = lookback - 1
    upper[valid_from:] = line[valid_from:] + mae
    lower[valid_from:] = line[valid_from:] - mae

    band_start = lookback * 2 - 2
    upper[:band_start] = np.nan
    lower[:band_start] = np.nan
    return upper, lower
